specs/PermissionManagement.md

```diff
@@ -0,0 +1,554 @@
+// to decide whether to show custom UI to the user.
+HRESULT SettingsComponent::OnPermissionRequested(
+    ICoreWebView2* sender, ICoreWebView2PermissionRequestedEventArgs* args)
+{
+    // Obtain a deferral for the event so that the CoreWebView2
+    // doesn't examine the properties we set on the event args until
+    // after we call the Complete method asynchronously later.
+    wil::com_ptr<ICoreWebView2Deferral> deferral;
+    CHECK_FAILURE(args->GetDeferral(&deferral));
+
+    // Do not persist state so that the PermissionRequested event is always
+    // raised and the app is in control of all permission requests.
+    auto extendedArgs = args.try_query<ICoreWebView2PermissionRequestedEventArgs3>();
+    if (extendedArgs)
+    {
+        CHECK_FAILURE(extendedArgs->put_ShouldPersist(FALSE));
+    }
+
+    // Do the rest asynchronously, to avoid calling dialog in an event handler.
+    m_appWindow->RunAsync([this, deferral, args]
+    {
+        COREWEBVIEW2_PERMISSION_KIND kind = COREWEBVIEW2_PERMISSION_KIND_UNKNOWN_PERMISSION;
+        BOOL userInitiated = FALSE;
+        wil::unique_cotaskmem_string uri;
+        CHECK_FAILURE(args->get_PermissionKind(&kind));
+        CHECK_FAILURE(args->get_IsUserInitiated(&userInitiated));
+        CHECK_FAILURE(args->get_Uri(&uri));
+
+        auto cachedKey = std::make_tuple(std::wstring(uri.get()), kind, userInitiated);
+        auto cachedPermission = m_cachedPermissions.find(cachedKey);
+        if (cachedPermission != m_cachedPermissions.end())
+        {
+            COREWEBVIEW2_PERMISSION_STATE state;
+            state =
+                (cachedPermission->second ? COREWEBVIEW2_PERMISSION_STATE_ALLOW
+                                           : COREWEBVIEW2_PERMISSION_STATE_DENY);
+            CHECK_FAILURE(args->put_State(state));
+        }
+        else
+        {
+            ShowPermissionDialog(uri, kind, userInitiated);
+        }
+        CHECK_FAILURE(deferral->Complete());
+    });
+    return S_OK;
+}
+```
+## SetPermission and GetNonDefaultPermissionCollection
+```c#
+// Gets the nondefault permission collection and updates a custom permission
+// management page.
+async void WebView_PermissionManager_DOMContentLoaded(object sender,
+    CoreWebView2DOMContentLoadedEventArgs arg)
+{
+    // Permission management APIs are only used on the app's permission
+    // management page.
+    if (webView.CoreWebView2.Source !=
+        "https://appassets.example/ScenarioPermissionManagement.html")
+    {
+        return;
+    }
+    // Get all the nondefault permissions and post them to the app's
+    // permission management page. The permission management page can present
+    // a list of custom permissions set for this profile and let the end user
+    // modify them.
+    IReadOnlyList<CoreWebView2PermissionSetting> permissionList =
+        await WebViewProfile.GetNonDefaultPermissionCollectionAsync();
```

Comment:
Is this `WebViewProfile` really `webView.CoreWebView2.Profile`? Put it into a variable?